LINKTA-344 Set Up Testing Environment for Integration Testing With Jest

ejwoods opened the pull request:
<!--# Pull Request Template -->

## Description

<!--
Please include a summary of the change and which issue is fixed. Include relevant motivation and context. List any dependencies that are required for this change.

Linking Issues:
 - To associate this pull request with a GitHub issue within the same repository, use the "Fixes #IssueNumber" syntax.
 - For external issues, use Markdown links: [Issue Name](Issue Link).
-->

Fixes # 344

This ticket implements basic configuration for integration testing using Jest.

- moved from a single `jest.config.ts` to using separate config files for unit and integration testing
  - I was hoping to utilize the [projects](https://jestjs.io/docs/configuration#projects-arraystring--projectconfig) config prop to handle multiple configurations within the same file, but was unable to figure out how to make it work with `nextJest`, which returns a configuration handler that does not accept `projects`. 
- updated package.json test scripts to run unit and integration tests via their separate configs
- updated jest-unit.yml to run new npm `test:unit` script
- added jest-integration.yml to create a GitHub Actions workflow for running integration tests on PRs to dev branch
- updated jest-unit.yml and jest-integration.yml to use `npm ci` for their 'Install dependencies' jobs. This is preferable to `npm install` in testing and automated environments because it always performs a clean install based on package-lock.json. More info in the [npm docs](https://docs.npmjs.com/cli/v8/commands/npm-ci).
- added `integration/` test directory with a placeholder test to confirm config, scripts, and GHA workflow were set up correctly

## Type of change

<!--Please delete options that are not relevant.-->

- [x] Other 🗒️ Testing

## How Has This Been Tested?

<!--Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Remove if not applicable. -->

- [x] `npm run test`

## Checklist:

<!--Before submitting your pull request, please review the following checklist:-->

- [x] I have performed a self-review of my own code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I have reviewed the [**Security Best Practices Document**](SECURITY_BEST_PRACTICES.md).
- [x] I have followed the [**Naming Conventions Guide**](NAMING_CONVENTIONS_GUIDE.md) for my changes.
- [x] I have ensured that my pull request title is descriptive.

## Additional Information:

<!--Any additional information, configuration, or data that might be necessary to reproduce the issue or feature.-->

ejwoods committed:
rename config to baseConfig in jest.config.ts

Rename config object in jest config to baseConfig. It will be used to create specific config objects for unit and integration testing.

ejwoods committed:
add unitConfig object to jest.config.ts

Running tests with this config will only run tests with the extensions .test.ts and .test.tsx within the unit test directory

ejwoods committed:
add integrationConfig object to jest.config.ts

Using this config will only run tests with the extensions .int-test.ts and .-int-test.tsx within the integration/ test directory.

ejwoods committed:
update Jest config and scripts for separate test environments

- Renamed jest.config.ts to jest.unit.config.ts to facilitate use of multiple configuration files. This is necessary because `nextJest` does not appear to support the `projects` config property, complicating use of a single config file. Maintaining separate config files will keep test environments distinct.
- Updated package.json test scripts to maintain functionality. All tests are currently passing.

ejwoods committed:
add jest.integration.config.ts file

Created jest.integration.config.ts file to set up testing environment for integration tests

ejwoods committed:
add integration test scripts to package.json

- Added package.json scripts to run integration test suites
- Updated `test` script to run integration tests after unit tests

ejwoods committed:
update jest-unit.yml to run unit test script

ejwoods committed:
add yml for integration testing workflow

Created jest-integration.yml to add integration testing workflow to GitHub Actions

ejwoods committed:
add integration test directory and placeholder test

Added __tests__/integration/ directory and placeholder.int-test.ts to ensure correct setup of:
- integration test configuration
- package.json test scripts
- integration test GitHub Actions workflow

Placeholder file should be removed when actual integration tests are added.

ejwoods committed:
fix: correct npm test scripts in test workflows

Added missing word 'run' to testing jobs in jest-unit.yml and jest-integration.yml

ejwoods committed:
fix: correct typo in jest-integration.yml

kinskiwu commented on .github/workflows/jest-integration.yml:
It seems we need to include setup steps related to Firestore to run our integration tests. Although I haven't looked deeply into this yet, a practical approach would be to use the Firestore Emulator Suite to allow the integration tests mimic real-world operations without the need to set up a test database.

Here are some resources I found while doing this code revivew, feel free to explore more on your own:
https://firebase.google.com/docs/emulator-suite
https://dev.to/ellioseven/a-next-js-firebase-tdd-environment-example-3ojp

ejwoods replied:
Thanks for pointing this Emulator Suite out - I hadn't come it across it yet. 

Since we don't have tickets for any specific tests yet, my thought for this PR is to set up the lowest level of configuration necessary. I have some tests in mind that I want to write (creating LINKTA-396 for them currently), but I haven't anticipated a need for this level of emulation for those tests specifically. I think it makes the most sense to add this configuration when tests that will use it are being written. What do you think?

kinskiwu replied:
Sounds like a good workaround. Do you mind updating the subject for this PR to indicate it is a basic setup? I will create a new ticket for the firebase integration.

kinskiwu replied:
Ticket created.

kinskiwu commented on __tests__/integration/placeholder.int-test.ts:
```diff
@@ -0,0 +1,5 @@
+describe('Integration test placeholder', () => {
```
To verify our testing environment setup, and following Martin's recommendation from the last standup about testing best practices, it might be helpful to implement a basic integration test related to Firestore. This will help ensure our setup is configured correctly.

kinskiwu commented on jest.integration.config.ts:
Since we have separate configs for unit and integration tests, directing jest to generate distinct coverage reports for each can boost performance and clarity. This can cut down on performance overhead and avoid mixing coverage data between test types. 

```
// For jest.integration.config.ts
coverageDirectory: 'coverage/integration',

// For jest.unit.config.ts
coverageDirectory: 'coverage/unit',
```

ejwoods replied:
Good tip, will implement this.

danielboisselle commented on .github/workflows/jest-integration.yml:
```diff
@@ -0,0 +1,28 @@
+          node-version: '20'
+        
+      - name: Install dependencies
+        run: npm install
```
`npm ci`

https://docs.npmjs.com/cli/v10/commands/npm-ci

ejwoods committed:
update npm install jobs for jest test workflows

Changed the 'Install dependencies' jobs in jest-unit.yml and jest-integration.yml to use 'npm ci' instead of 'npm install'. This performs a clean install of dependencies based on package-lock.json while 'npm install' will install from a cache if available.

ejwoods committed:
update test configs to separate coverage directories

Specify separate coverage directories for unit and integration tests in jest.unit.config.ts and jest.integration.config.ts.
